File: frontend/components/forms/dimensionUnits.ts
/**
 * CENTIMETRES ↔ INCHES FOR THE RECORD FORMS — the pure half, so a judgement about a stored number
 * is exercised by a test rather than only by somebody looking at a screen.
 *
 * `ToolDocumentation.height` and `.width` are the CENTIMETRE boxes and `heightInches` /
 * `breadthInches` are their inch partners; `lengthInches` is standalone and has no partner. Typing
 * in either box of a pair fills the other, and this module is the whole of the arithmetic and the
 * whole of the rule about WHEN a partner may be written.
 *
 * ── WHY IT IS NOT INSIDE `ToolForm.tsx` ─────────────────────────────────────────────────────────
 * There is no React renderer in this repository's devDependencies — Playwright is the whole of it —
 * so a decision written inside JSX can never be driven by a test. The same split, for the same
 * reason, as `components/forms/measurementMethods.ts`, `components/media/gridProposal.ts` and
 * `components/ui/selectFilter.ts`. `frontend/e2e/dimension-units-unit.spec.ts` drives every row of
 * the table below.
 *
 * ── WHY THE ARITHMETIC IS SPELLED THIS PEDANTICALLY ─────────────────────────────────────────────
 * The identical function exists on the handset (`android/.../ui/DimensionUnits.kt`) and the two
 * clients write into the same `Decimal(10, 2)` columns. If they disagree by a hundredth, a tool
 * edited on a laptop and then on a phone changes its own width on every save, silently, for ever.
 * So every step below is chosen to be bit-identical between binary64 in V8 and `Double` on the JVM:
 *
 *  1. ONE GRAMMAR, not each language's own parser. `Number("0x1A")` is 26 while `"0x1A".toDouble()`
 *     throws; `"1.5f".toDouble()` is 1.5 while `Number("1.5f")` is NaN; both accept `"Infinity"`.
 *     A regex both languages implement identically is the only shape the four clients cannot drift
 *     apart on.
 *  2. ONE ROUNDING, expressed as `floor(x + 0.5)`. `kotlin.math.round` is half-away-from-zero and
 *     the JVM's `Math.round` is `floor(x + 0.5)`; they differ for negatives, and a latent
 *     divergence is not made safe by the grammar currently refusing a minus sign.
 *  3. THE OUTPUT NEVER PASSES THROUGH EITHER LANGUAGE'S FLOAT-TO-STRING. JavaScript prints `3` for
 *     an integral 3.0 and Kotlin prints `3.0`. {@link renderScaled} builds the text from the scaled
 *     INTEGER instead.
 *  4. ASSOCIATION ORDER IS LOAD-BEARING. `v * 2.54 * 100` must not be rewritten as `v * 254`, and
 *     `v / 2.54 * 100` must not become `v * 0.3937007874015748`. Each rewrite moves the last ulp and
 *     can flip a hundredth.
 *
 * ── AND WHY IT IS NOT THE BACKEND'S `_inches_to_cm` ─────────────────────────────────────────────
 * `backend/app/services/design_workshops.py` has an `_inches_to_cm` that reads
 * `round(float(v) * 2.54, 2)` — Python's BANKER'S rounding, a different halfway rule from the one
 * here. The two live on opposite sides of the wire, are applied to different columns (that one
 * carries a tool's inches into a design-workshop stage box; this one pairs two columns of the tool
 * record itself), and must NOT be unified: doing so would change every workshop carry figure that
 * lands on a halfway value. The cross-reference is written in both directions on purpose.
 */

/**
 * EXACT. The inch has been defined as 25.4 mm since the 1959 international yard-and-pound
 * agreement, so this is not a measured constant with an error bar and never "2.5 for readability".
 * Named identically to the backend's `design_workshops._CM_PER_INCH` and to Android's
 * `DimensionUnits.CM_PER_INCH`.
 */
export const CM_PER_INCH = 2.54;

/**
 * The one grammar — see the header. No sign, no exponent, no hex, no thousands separator, no
 * trailing type suffix; leading and trailing spaces and tabs only.
 *
 * `1.` AND `.5` ARE BOTH ADMITTED, DELIBERATELY — AND THE BROWSER'S OWN BOXES NEVER PRODUCE EITHER.
 * They are what a person mid-decimal has actually typed, and on the handset that is literally what
 * arrives: `TextInput(… keyboardType = KeyboardType.Decimal)` hands `onValueChange` the raw string,
 * so `propagateDimension` really does see `"1."`. In this client every paired box is a native
 * `<input type="number">`, whose `.value` is `""` for anything the HTML floating-point grammar
 * refuses — and `"1."` is refused, because that grammar requires a digit after the point. So the
 * admission here is for the TWIN and for callers that are not an input at all (the two measurement
 * routes hand {@link propagate} a formatted reading directly), not for anything a browser keystroke
 * can reach. Keep it: a grammar the four clients do not share is how they come to disagree by a
 * hundredth. See {@link propagate}'s rule 4 for what the browser does instead.
 */
const NUMERIC_RE = /^[ \t]*([0-9]+\.?[0-9]*|\.[0-9]+)[ \t]*$/;

/** `99,999,999.99`, scaled by 100 — the largest value a `Decimal(10, 2)` column can hold. */
const DECIMAL_10_2_MAX_SCALED = 9999999999;

/**
 * The number in a dimension box, or null where the text cannot be one.
 *
 * Non-negative by construction: the grammar admits no minus sign, which is the same bound the boxes
 * carry as `min={0}` and the schemas carry as `ge=0`.
 *
 * THE `isFinite` GUARD IS REACHABLE, contrary to what a first reading of the grammar suggests: a
 * four-hundred-digit integer matches {@link NUMERIC_RE} perfectly and `Number()` answers `Infinity`
 * for it, in both languages. Without the guard `Infinity * 2.54 * 100` would pass the
 * `Decimal(10, 2)` ceiling test — `Infinity > 9999999999` is true, so it is refused there too — but
 * only by accident of the comparison, and a value that is not a number must be refused for being one
 * rather than for being large. The Kotlin twin carries the same check for the same reason.
 */
export function parseDimension(text: string): number | null {
  if (!NUMERIC_RE.test(text)) return null;
  const value = Number(text.trim());
  return Number.isFinite(value) ? value : null;
}

/**
 * Hundredths, from the scaled integer — never from `Number.prototype.toString`.
 *
 * `254 → "254"`, `2540 → "25.4"`, `318 → "3.18"`. Trailing zeros are dropped so the text reads the
 * way a person would have typed it, and so an inch value that converts back to a whole number of
 * centimetres does not print `"254.00"` into a box whose partner said `"100"`.
 */
function renderScaled(scaled: number): string {
  const whole = Math.floor(scaled / 100);
  const frac = scaled % 100;
  if (frac === 0) return String(whole);
  if (frac % 10 === 0) return `${whole}.${frac / 10}`;
  return `${whole}.${frac < 10 ? `0${frac}` : frac}`;
}

/** Centimetres, two decimals, from an inch box's text — or null where nothing may be written. */
export function cmTextFromInches(text: string): string | null {
  const value = parseDimension(text);
  if (value === null) return null;
  const scaled = Math.floor(value * CM_PER_INCH * 100 + 0.5);
  if (scaled > DECIMAL_10_2_MAX_SCALED) return null;
  return renderScaled(scaled);
}

/** Inches, two decimals, from a centimetre box's text — or null where nothing may be written. */
export function inchesTextFromCm(text: string): string | null {
  const value = parseDimension(text);
  if (value === null) return null;
  const scaled = Math.floor((value / CM_PER_INCH) * 100 + 0.5);
  if (scaled > DECIMAL_10_2_MAX_SCALED) return null;
  return renderScaled(scaled);
}

/**
 * WRITE THE PARTNER BOX — the four rules that keep this pairing from eating a stored number.
 *
 * 1. **ONE-DIRECTIONAL.** This is only ever called from the change handler of the box a person is
 *    typing in, and it only ever writes the OTHER box. It must never be reached from an effect
 *    watching both values: a watcher fires for the value it just wrote, and `1 cm → 0.39 in →
 *    0.99 cm` is how a round trip eats a hundredth per keystroke. React's `setState` does not
 *    re-invoke the target control's own `onChange`, so the natural shape is already correct — the
 *    rule is here to stop somebody "simplifying" it into a `useEffect`.
 * 2. **CLEARING PROPAGATES, AND IT IS THE FIRST BRANCH.** An empty box means "no value", and
 *    leaving a stale converted number standing beside it is a lie about a measurement. This is the
 *    one input that fails the grammar and still writes the partner, which is why it is tested
 *    before the parse rather than falling out of it.
 * 3. **ANYTHING ELSE THAT IS NOT A NUMBER WRITES NOTHING.** `"-"`, `"1.2.3"`, `"abc"`, `"1e3"` and
 *    a value past `Decimal(10, 2)` leave the partner exactly as it is.
 * 4. **A PARTIALLY TYPED DECIMAL IS A NUMBER, NOT A MISTAKE** — where the caller can hand one over.
 *    `"1."` parses and converts, so the partner tracks the keystrokes instead of blanking and
 *    refilling, and `"1.2."` (not a number) holds `3.05` for exactly the one keystroke it takes to
 *    finish typing rather than claiming the designer had said there is no width.
 *
 * ── AND RULE 4 DOES NOT DESCRIBE THIS CLIENT'S KEYSTROKES, WHICH IS WORTH KNOWING BEFORE YOU
 *    DEBUG THE CENTIMETRE BOX BLINKING ──────────────────────────────────────────────────────────
 * Every paired box on `ToolForm` is `<input type="number">`, and a native number input reports `""`
 * for text the HTML floating-point grammar refuses — `"1."` among it. So the keystroke that leaves
 * `"3."` in the Height (inches) box arrives here as the EMPTY string, takes rule 2, and blanks the
 * centimetre partner; the next digit refills it. It is visible, it is not data loss by itself, and
 * it has one sharp edge: a save pressed at exactly that instant sends `null` for BOTH columns of the
 * pair, because the state really is empty while the box still draws `3.` (the browser keeps the
 * rejected text, and React skips a DOM write it believes is already `""`).
 *
 * The rule is kept as it is because the grammar is a four-client contract (see {@link NUMERIC_RE}):
 * the handset's Compose field DOES hand over `"1."`, and so do the measurement routes, which call
 * {@link propagate} with a formatted reading and never touch an input. Making rule 4 reachable in the
 * browser would mean `type="text" inputMode="decimal"` on all four boxes, which trades this blink for
 * the loss of `min={0}` refusal at the control — a worse bargain on columns the schema bounds `ge=0`.
 */
export function propagate(
  text: string,
  convert: (value: string) => string | null,
  setPartner: (value: string) => void
): void {
  if (text.trim() === "") {
    setPartner("");
    return;
  }
  const next = convert(text);
  if (next !== null) setPartner(next);
}

File: android/app/src/main/java/com/designprototype/workshop/ui/DimensionUnits.kt
package com.designprototype.workshop.ui

import kotlin.math.floor

/*
 * CENTIMETRES ↔ INCHES FOR THE RECORD FORMS — the Kotlin half of
 * `frontend/components/forms/dimensionUnits.ts`, arithmetic for arithmetic.
 *
 * ── WHAT THIS IS FOR ──────────────────────────────────────────────────────────────────────────
 *
 * `ToolDocumentation` carries `height` and `width` as bare `Decimal(10,2)` columns and
 * `heightInches` / `breadthInches` / `lengthInches` beside them. Since 2026-09-15 the first two are
 * CENTIMETRES and each pairs 1:1 with one of the inch boxes — Height (cm) ↔ Height (inches), Width
 * (cm) ↔ Breadth (inches) — so a designer fills either box of a pair and the other follows. Length
 * (inches) keeps no partner and gets no centimetre column.
 *
 * ── WHY IT LIVES IN `ui/` AND NOT IN `MainActivity.kt` ────────────────────────────────────────
 *
 * The same reason `ui/RecordPickers.kt` gives for `craftsChangeClearsArtisans`: this is where a JVM
 * unit test can reach it. `DimensionUnitsTest` drives every row of the cross-surface table, and a
 * judgement written inside a `@Composable` in `MainActivity.kt` could not be exercised at all —
 * there is no `ui-test-junit4` and no Robolectric in `app/build.gradle.kts`.
 *
 * ── WHY IT IS A SECOND IMPLEMENTATION AND NOT A CALL TO THE SERVER'S ─────────────────────────
 *
 * `backend/app/services/design_workshops._inches_to_cm` converts too, and the two MUST NOT be
 * unified. That one is `round(float(v) * 2.54, 2)` — Python's round-half-to-EVEN — and it converts a
 * DIFFERENT number for a different reader: a tool's inch height into a design-workshop stage's
 * `heightCm` box, on the server, after the record is saved. This one is round-half-UP and runs
 * against the keystrokes of somebody typing into a record form. Unifying them would change every
 * workshop carry figure that lands on a halfway value, silently, in documents that have already been
 * printed. Two jobs, two rules, and a comment on each naming the other.
 *
 * `ui/designworkshop/DwReferenceField.kt`'s "NO UNIT CONVERSION HAPPENS HERE, AND NONE MAY EVER BE
 * ADDED" is a third thing again and is unaffected: that rule is about the STAGE side, where the
 * server has already converted and a second conversion would double it.
 *
 * ── THE PARITY RULES, WHICH ARE WHY THIS FILE LOOKS OVER-ENGINEERED ──────────────────────────
 *
 * Four clients implement this — two browsers and two handsets — and they must agree to the last
 * hundredth on the same typed string, because the number they write is the number a ministry reads.
 * Three properties carry that guarantee, and every one of them is a line somebody could "tidy":
 *
 *   1. ONE GRAMMAR, ASSERTED BY REGEX, because the two languages' own parsers disagree at the
 *      edges. `Number("0x1A")` is 26 while `"0x1A".toDouble()` throws; `"1.5f".toDouble()` is 1.5
 *      while `Number("1.5f")` is NaN; both accept "Infinity" and "NaN". A regex both languages
 *      implement identically is the only shape that cannot drift.
 *   2. ONE ROUNDING, `floor(x + 0.5)`, spelled out rather than delegated. `kotlin.math.round` is
 *      half-away-from-zero and JVM `Math.round` is `floor(x+0.5)`; the two differ for negatives, and
 *      leaving that latent in a file whose grammar happens to admit no sign today is how it becomes
 *      a defect the day the grammar changes.
 *   3. NO FLOAT-TO-STRING. [renderScaled] builds the text out of the scaled INTEGER, because JS
 *      prints "3" for an integral 3.0 where Kotlin prints "3.0", and the two disagree again on large
 *      magnitudes. The digits a designer sees are assembled, never formatted.
 *
 * Association order is load-bearing too. `v * 2.54 * 100.0` must not become `v * 254.0`, and
 * `v / 2.54 * 100.0` must not become `v * (100.0 / 2.54)`: each rewrite moves the last ulp and can
 * flip a hundredth. Kotlin's `*` and `/` are left-associative and so are ECMAScript's, so the two
 * expressions below are the same sequence of correctly-rounded binary64 operations on both sides.
 */

/**
 * Centimetres in one inch, EXACT.
 *
 * The inch has been defined as 25.4 mm since the 1959 international yard-and-pound agreement, so
 * this is a definition rather than a measurement and there is no "2.5 for readability" available.
 * Named to match `dimensionUnits.ts`'s `CM_PER_INCH` and the backend's
 * `design_workshops._CM_PER_INCH`, so a grep for the constant finds all three.
 */
const val CM_PER_INCH: Double = 2.54

/**
 * The one grammar all four clients admit: optional space/tab padding around an unsigned decimal.
 *
 * `1`, `1.`, `.5`, `0.50` are numbers. `-1`, `1e3`, `0x1A`, `1.5f`, `Infinity`, `NaN`, `1,5` and
 * `1 000` are not, on every client, whatever the local parser would have made of them. No sign is
 * admitted at all, which is what makes "non-negative" a property of the grammar rather than a check
 * somebody has to remember to write.
 */
private val NUMERIC_RE = Regex("""^[ \t]*([0-9]+\.?[0-9]*|\.[0-9]+)[ \t]*$""")

/** 99,999,999.99 × 100 — the largest value `Decimal(10,2)` can hold, in hundredths. */
private const val DECIMAL_10_2_MAX_SCALED: Long = 9_999_999_999L

/**
 * [text] as a finite non-negative number, or null when it is not one.
 *
 * NULL IS NOT AN ERROR; it is the answer for every string a partner box must not be written from —
 * an empty box, a lone "-", a half-typed "1.2.", a word. The caller decides what to do with it, and
 * for the record forms the answer is "write nothing", which is the only answer that cannot put a
 * number the designer never typed into a column a ministry reads.
 *
 * THE `isFinite` CHECK IS REACHABLE, contrary to what a first reading of the grammar suggests: a
 * four-hundred-digit integer matches [NUMERIC_RE] perfectly and `Double.parseDouble` answers
 * `Infinity` for it, exactly as JavaScript's `Number` does. Keeping the check is what makes this a
 * total function rather than one that returns an infinity for somebody to multiply.
 */
fun parseDimension(text: String): Double? {
    if (!NUMERIC_RE.matches(text)) return null
    val value = text.trim().toDoubleOrNull() ?: return null
    if (!value.isFinite()) return null
    return value
}

/**
 * [text] read as inches, in hundredths of a centimetre, or null when there is nothing to convert.
 *
 * ONE ROUNDING, NOT TWO. Converting and then formatting from a re-derived double would round twice
 * and can move the last hundredth, which is the whole reason the scaled integer — not the converted
 * double — is what leaves this function.
 */
private fun cmScaledFromInches(text: String): Long? {
    val value = parseDimension(text) ?: return null
    val scaled = floor(value * CM_PER_INCH * 100.0 + 0.5)
    if (scaled > DECIMAL_10_2_MAX_SCALED) return null
    return scaled.toLong()
}

/** [text] read as centimetres, in hundredths of an inch. See [cmScaledFromInches]. */
private fun inchesScaledFromCm(text: String): Long? {
    val value = parseDimension(text) ?: return null
    val scaled = floor(value / CM_PER_INCH * 100.0 + 0.5)
    if (scaled > DECIMAL_10_2_MAX_SCALED) return null
    return scaled.toLong()
}

/**
 * Hundredths as the shortest honest decimal: `254` → "2.54", `2540` → "25.4", `25400` → "254".
 *
 * BUILT FROM THE INTEGER, never from `Double.toString()`. The trailing zero is dropped because that
 * is what a person typing "10" and seeing "25.4" expects, and because the browser's twin drops it —
 * the two clients must produce the same STRING, not merely the same quantity, or a designer moving
 * between them reads a record as edited when nothing changed.
 */
private fun renderScaled(scaled: Long): String {
    val whole = scaled / 100
    val frac = scaled % 100
    return when {
        frac == 0L -> whole.toString()
        frac % 10L == 0L -> "$whole.${frac / 10}"
        else -> "$whole." + frac.toString().padStart(2, '0')
    }
}

/** Inches as centimetres, rounded to two decimals, or null when [text] is not a number. */
fun cmTextFromInches(text: String): String? = cmScaledFromInches(text)?.let(::renderScaled)

/** Centimetres as inches, rounded to two decimals, or null when [text] is not a number. */
fun inchesTextFromCm(text: String): String? = inchesScaledFromCm(text)?.let(::renderScaled)

/**
 * Write [text]'s partner box, by the three rules that stop this pairing corrupting data.
 *
 * Called from the change handler of the box the designer is ACTUALLY TYPING IN, and from nowhere
 * else. That is not a style preference, it is the whole safety property:
 *
 *   ONE-DIRECTIONAL PER KEYSTROKE. The box being written must never turn round and re-derive its
 *   source. `1 cm → 0.39 in → 0.99 cm` is how a round trip eats a value, and a watcher — a
 *   `LaunchedEffect(height, heightInches)` — fires for BOTH values and is exactly that round trip.
 *   Compose state assignment does not re-invoke the target control's `onValueChange`, so writing the
 *   partner from inside the source's own handler is already the correct shape; there is nothing to
 *   add and there is a great deal to avoid adding.
 *
 *   CLEARING PROPAGATES. An empty box means "no value", and leaving a stale converted number
 *   standing beside it is a lie the next reader has no way to detect. This is the ONE case where a
 *   string that is not a number still writes the partner, which is why it is its own branch above
 *   the parse rather than a special case inside it.
 *
 *   A NON-NUMBER WRITES NOTHING. A partially typed decimal is a number, not a mistake: "1." parses
 *   and converts, so the partner tracks the keystrokes instead of blanking and refilling. Only a
 *   string that cannot be a number at all leaves the partner alone — and only an EMPTY box clears
 *   it, because empty is the one input that means "no value". So `1` → `1.` → `1.5` gives the
 *   partner `2.54` → `2.54` → `3.81` with no flicker, and `1.2.` holds `3.05` for the one keystroke
 *   it takes to finish typing. Stale for a moment beats blank-and-back, because blank is a claim.
 *
 * NEVER CALLED ON LOAD. Opening an existing record must not rewrite either box: historic rows
 * genuinely hold unrelated numbers in `height` and `heightInches` — the schema comment and the tool
 * form's own note both say so — and a load-time conversion would destroy them on the next save.
 */
fun propagateDimension(text: String, convert: (String) -> String?, setPartner: (String) -> Unit) {
    if (text.isBlank()) {
        setPartner("")
        return
    }
    val next = convert(text) ?: return
    setPartner(next)
}
